File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/configuration/sdk_environment.dart
import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/native/dart_bridge_dev_config.dart';

/// SDK Environment mode - determines how data is handled
enum SDKEnvironment {
  /// Development/testing mode - may use local data, verbose logging
  development,

  /// Staging mode - testing with real services
  staging,

  /// Production mode - live environment
  production,
}

extension SDKEnvironmentExtension on SDKEnvironment {
  /// Human-readable description
  String get description {
    switch (this) {
      case SDKEnvironment.development:
        return 'Development Environment';
      case SDKEnvironment.staging:
        return 'Staging Environment';
      case SDKEnvironment.production:
        return 'Production Environment';
    }
  }

  /// Check if this is a production environment
  bool get isProduction => this == SDKEnvironment.production;

  /// Check if this is a testing environment
  bool get isTesting =>
      this == SDKEnvironment.development || this == SDKEnvironment.staging;

  /// Check if this environment requires a valid backend URL
  bool get requiresBackendURL => this != SDKEnvironment.development;

  // MARK: - Build Configuration Validation

  /// Check if the current build configuration is compatible with this environment
  /// Production environment is only allowed in Release builds
  bool get isCompatibleWithCurrentBuild {
    switch (this) {
      case SDKEnvironment.development:
      case SDKEnvironment.staging:
        return true;
      case SDKEnvironment.production:
        // In Dart/Flutter, we use kDebugMode or assert() to check debug mode
        // Since there's no #if DEBUG in Dart, we check via assert
        bool isDebug = false;
        assert(() {
          isDebug = true;
          return true;
        }());
        return !isDebug;
    }
  }

  /// Returns true if we're running in a DEBUG build
  static bool get isDebugBuild {
    bool isDebug = false;
    assert(() {
      isDebug = true;
      return true;
    }());
    return isDebug;
  }

  // MARK: - Environment-Specific Settings

  /// Determine logging verbosity based on environment
  LogLevel get defaultLogLevel {
    switch (this) {
      case SDKEnvironment.development:
        return LogLevel.debug;
      case SDKEnvironment.staging:
        return LogLevel.info;
      case SDKEnvironment.production:
        return LogLevel.warning;
    }
  }

  /// Should send telemetry data (production only)
  bool get shouldSendTelemetry => this == SDKEnvironment.production;

  /// Should use mock data sources (development only)
  bool get useMockData => this == SDKEnvironment.development;

  /// Should sync with backend (non-development)
  bool get shouldSyncWithBackend => this != SDKEnvironment.development;

  /// Requires API authentication (non-development)
  bool get requiresAuthentication => this != SDKEnvironment.development;
}

/// Supabase configuration
class SupabaseConfig {
  final Uri projectURL;
  final String anonKey;

  SupabaseConfig({
    required this.projectURL,
    required this.anonKey,
  });

  /// Get configuration for environment
  /// 
  /// For development mode, reads from C++ dev config (development_config.cpp).
  /// This ensures credentials are stored in a single git-ignored location.
  static SupabaseConfig? configuration(SDKEnvironment environment) {
    switch (environment) {
      case SDKEnvironment.development:
        // Read from C++ dev config - credentials stored in development_config.cpp (git-ignored)
        final supabaseUrl = DartBridgeDevConfig.supabaseURL;
        final supabaseKey = DartBridgeDevConfig.supabaseKey;
        
        if (supabaseUrl == null || supabaseUrl.isEmpty ||
            supabaseKey == null || supabaseKey.isEmpty) {
          // Dev config not available - this is expected if development_config.cpp 
          // hasn't been filled in. Telemetry will be disabled in dev mode.
          return null;
        }
        
        return SupabaseConfig(
          projectURL: Uri.parse(supabaseUrl),
          anonKey: supabaseKey,
        );
      case SDKEnvironment.staging:
      case SDKEnvironment.production:
        return null;
    }
  }
}

/// SDK initialization parameters
class SDKInitParams {
  /// API key for authentication
  final String apiKey;

  /// Base URL for API requests
  final Uri baseURL;

  /// Environment mode
  final SDKEnvironment environment;

  /// Supabase configuration (for analytics in dev mode)
  SupabaseConfig? get supabaseConfig =>
      SupabaseConfig.configuration(environment);

  SDKInitParams({
    required this.apiKey,
    required this.baseURL,
    this.environment = SDKEnvironment.production,
  });

  /// Create from string URL
  factory SDKInitParams.fromString({
    required String apiKey,
    required String baseURL,
    SDKEnvironment environment = SDKEnvironment.production,
  }) {
    final uri = Uri.tryParse(baseURL);
    if (uri == null) {
      throw ArgumentError('Invalid base URL: $baseURL');
    }
    return SDKInitParams(
      apiKey: apiKey,
      baseURL: uri,
      environment: environment,
    );
  }

  /// Create development mode parameters
  /// Uses Supabase for analytics, no authentication required
  /// Matches iOS SDKInitParams(forDevelopmentWithAPIKey:)
  factory SDKInitParams.forDevelopment({String apiKey = ''}) {
    final supabaseConfig =
        SupabaseConfig.configuration(SDKEnvironment.development);
    return SDKInitParams(
      apiKey: apiKey,
      baseURL: supabaseConfig?.projectURL ?? Uri.parse('http://localhost'),
      environment: SDKEnvironment.development,
    );
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Configuration/SDKEnvironment.swift
import CRACommons
import Foundation

/// SDK Environment mode - determines how data is handled
public enum SDKEnvironment: String, CaseIterable, Sendable {
    /// Development/testing mode - may use local data, verbose logging
    case development

    /// Staging mode - testing with real services
    case staging

    /// Production mode - live environment
    case production

    // MARK: - C++ Bridge

    /// Convert to C++ environment type for cross-platform consistency
    var cEnvironment: rac_environment_t {
        switch self {
        case .development: return RAC_ENV_DEVELOPMENT
        case .staging: return RAC_ENV_STAGING
        case .production: return RAC_ENV_PRODUCTION
        }
    }

    /// Human-readable description
    public var description: String {
        switch self {
        case .development:
            return "Development Environment"
        case .staging:
            return "Staging Environment"
        case .production:
            return "Production Environment"
        }
    }

    /// Check if this is a production environment (uses C++)
    public var isProduction: Bool {
        rac_env_is_production(cEnvironment)
    }

    /// Check if this is a testing environment (uses C++)
    public var isTesting: Bool {
        rac_env_is_testing(cEnvironment)
    }

    /// Check if this environment requires a valid backend URL (uses C++)
    public var requiresBackendURL: Bool {
        rac_env_requires_backend_url(cEnvironment)
    }

    // MARK: - Build Configuration Validation

    /// Check if the current build configuration is compatible with this environment
    /// Production environment is only allowed in Release builds
    public var isCompatibleWithCurrentBuild: Bool {
        switch self {
        case .development, .staging:
            return true
        case .production:
            #if DEBUG
            return false
            #else
            return true
            #endif
        }
    }

    /// Returns true if we're running in a DEBUG build
    public static var isDebugBuild: Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }

    // MARK: - Environment-Specific Settings

    /// Determine logging verbosity based on environment
    public var defaultLogLevel: LogLevel {
        switch self {
        case .development: return .debug
        case .staging: return .info
        case .production: return .warning
        }
    }

    /// Should send telemetry data (production only) - uses C++
    public var shouldSendTelemetry: Bool {
        rac_env_should_send_telemetry(cEnvironment)
    }

    /// Should use mock data sources (development only)
    public var useMockData: Bool {
        self == .development // Keep simple - no C++ equivalent
    }

    /// Should sync with backend (non-development) - uses C++
    public var shouldSyncWithBackend: Bool {
        rac_env_should_sync_with_backend(cEnvironment)
    }

    /// Requires API authentication (non-development) - uses C++
    public var requiresAuthentication: Bool {
        rac_env_requires_auth(cEnvironment)
    }
}

/// SDK initialization parameters
public struct SDKInitParams {
    /// API key for authentication
    public let apiKey: String

    /// Base URL for API requests
    /// - Required for staging and production environments
    /// - Optional for development (uses placeholder if not provided)
    public let baseURL: URL

    /// Environment mode (development/staging/production)
    public let environment: SDKEnvironment

    // MARK: - Default Development URL

    /// Placeholder URL used for development when no URL is provided.
    /// Development mode uses local analytics, so this is just a placeholder.
    private static let developmentPlaceholderURL: URL = {
        guard let url = URL(string: "https://dev.runanywhere.local") else {
            fatalError("Invalid hardcoded development URL")
        }
        return url
    }()

    // MARK: - Initializers

    /// Create initialization parameters for staging or production
    /// - Parameters:
    ///   - apiKey: Your RunAnywhere API key (required)
    ///   - baseURL: Base URL for API requests (required, must be valid HTTPS URL)
    ///   - environment: Environment mode (default: production)
    /// - Throws: SDKError if validation fails
    public init(
        apiKey: String,
        baseURL: URL,
        environment: SDKEnvironment = .production
    ) throws {
        self.apiKey = apiKey
        self.baseURL = baseURL
        self.environment = environment

        // Validate based on environment
        try Self.validate(apiKey: apiKey, baseURL: baseURL, environment: environment)
    }

    /// Convenience initializer with string URL for staging or production
    /// - Parameters:
    ///   - apiKey: Your RunAnywhere API key
    ///   - baseURL: Base URL string for API requests
    ///   - environment: Environment mode (default: production)
    /// - Throws: SDKError if URL is invalid or validation fails
    public init(
        apiKey: String,
        baseURL: String,
        environment: SDKEnvironment = .production
    ) throws {
        guard let url = URL(string: baseURL) else {
            throw SDKError.general(.validationFailed, "Invalid base URL format: \(baseURL)")
        }
        try self.init(apiKey: apiKey, baseURL: url, environment: environment)
    }

    /// Convenience initializer for development mode (no URL required)
    /// - Parameter apiKey: Optional API key (not required for development)
    /// - Note: Development mode uses Supabase internally for dev analytics
    public init(forDevelopmentWithAPIKey apiKey: String = "") {
        self.apiKey = apiKey
        self.baseURL = Self.developmentPlaceholderURL
        self.environment = .development
    }

    // MARK: - Validation (Uses C++ for cross-platform consistency)

    /// Validate initialization parameters based on environment
    /// Uses C++ validation logic for cross-platform consistency.
    /// - Parameters:
    ///   - apiKey: The API key to validate
    ///   - baseURL: The base URL to validate
    ///   - environment: The target environment
    /// - Throws: SDKError if validation fails
    private static func validate(
        apiKey: String,
        baseURL: URL,
        environment: SDKEnvironment
    ) throws {
        let logger = SDKLogger(category: "SDKInitParams")

        // Note: We allow any environment in DEBUG builds to support developer testing
        // with custom backends. The environment parameter is informational for
        // logging and behavior configuration, not a security boundary.

        // Call C++ validation for API key and URL
        let cEnv = environment.cEnvironment

        // Validate API key via C++
        let apiKeyResult = apiKey.withCString { ptr in
            rac_validate_api_key(ptr, cEnv)
        }
        if apiKeyResult != RAC_VALIDATION_OK {
            let message = String(cString: rac_validation_error_message(apiKeyResult))
            switch apiKeyResult {
            case RAC_VALIDATION_API_KEY_REQUIRED:
                throw SDKError.general(.invalidAPIKey, "\(message) for \(environment.description)")
            case RAC_VALIDATION_API_KEY_TOO_SHORT:
                throw SDKError.general(.invalidAPIKey, message)
            default:
                throw SDKError.general(.validationFailed, message)
            }
        }

        // Validate URL via C++
        let urlResult = baseURL.absoluteString.withCString { ptr in
            rac_validate_base_url(ptr, cEnv)
        }
        if urlResult != RAC_VALIDATION_OK {
            let message = String(cString: rac_validation_error_message(urlResult))
            throw SDKError.general(.validationFailed, message)
        }

        // Log warnings for staging HTTP (C++ validates but doesn't warn)
        if environment == .staging, baseURL.scheme?.lowercased() == "http" {
            logger.warning("Using HTTP for staging environment. Consider using HTTPS for security.")
        }

        // Log warnings for staging localhost (C++ validates but doesn't warn)
        if environment == .staging, let host = baseURL.host?.lowercased() {
            if host.contains("localhost") || host.contains("127.0.0.1") ||
               host.contains("example.com") || host.contains(".local") {
                logger.warning("Staging environment using local/example URL: \(host)")
            }
        }

        logger.info("URL validated for \(environment.description): \(baseURL.absoluteString)")
    }
}
